tracing/src/main/java/com/palantir/remoting1/tracing/Tracer.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting1.tracing;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Sets;
+import java.util.Set;
+
+/**
+ * The singleton entry point for handling Zipkin-style traces and spans. Provides functionality for starting and
+ * completing spans, and for subscribing observers to span completion events.
+ * <p>
+ * This class is thread-safe.
+ */
+public final class Tracer {
+
+    private Tracer() {}
+
+    // Thread-safe since thread-local
+    private static final ThreadLocal<Trace> currentTrace = new ThreadLocal<Trace>() {
+        @Override
+        protected Trace initialValue() {
+            return new Trace(sampler.sample(), Traces.randomId());
+        }
+    };
+
+    // Thread-safe set implementation
+    private static final Set<SpanObserver> observers = Sets.newConcurrentHashSet();
+
+    // Thread-safe since stateless
+    private static TraceSampler sampler = AlwaysSampler.INSTANCE;
+
+    /**
+     * Initializes the current thread's trace, erasing any previously accrued open spans. The new trace is {@link
+     * Trace#isObservable observable} iff the given flag is true, or, iff {@code isObservable} is absent, if the {@link
+     * #setSampler configured sampler} returns true.
+     */
+    public static void initTrace(Optional<Boolean> isObservable, String traceId) {
+        boolean observable = isObservable.or(sampler.sample());
+        currentTrace.set(new Trace(observable, traceId));
+    }
+
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation and parent span.
+     */
+    public static OpenSpan startSpan(String operation, String parentSpanId) {
+        Preconditions.checkState(currentTrace.get().isEmpty(),
+                "Cannot start a span with explicit parent if the current thread's trace is non-empty");
```

Comment:
not a big fan of that since then one initTrace would also add a span and the other one would not?